#include "shared_screen.h"
#include <ui_shared_screen.h>
#include <QDateTime>
#include <QDockWidget>
#include <QHBoxLayout>
#include <QIcon>
#include <QLabel>
#include <QListWidget>
#include <QMenu>
#include <QMessageBox>
#include <QPushButton>
#include <QRandomGenerator>
#include <QShortcut>
#include <QTextBrowser>
#include <QTimer>
#include <QVBoxLayout>
#include <QKeyEvent>
#include <QDebug>

#include "signaling/TcpSignalingClient.hpp"
#include "rtc/PeerConnectionManager.hpp"


shared_screen::shared_screen(QWidget *parent)
    : QMainWindow(parent),
      ui(new Ui::shared_screen)
{
    ui->setupUi(this);

    // ================== P2P åˆå§‹åŒ– ==================
    qDebug() << "=== P2P Start ===";
    m_isCaller = true;  // è¿™å°æœºå™¨ä½œä¸º Callerï¼›å¦ä¸€å°ä½ å¯ä»¥æ”¹æˆ false
    m_signaling = new TcpSignalingClient(this);
    // å‡è®¾ä¿¡ä»¤æœåŠ¡å™¨åœ¨æœ¬æœº 127.0.0.1:12345
    m_signaling->connectToServer("127.0.0.1", 12345);
    qDebug() << "connectToServer done....";
    m_pcm = new PeerConnectionManager(m_signaling, m_isCaller, this);
    qDebug() << "new PeerConnectionManager done....";
    m_pcm->start();
    qDebug() << "m_pcm->start() done....";

    // ... P2P connect ä¿¡å·æ§½ ...
    connect(m_pcm, &PeerConnectionManager::connected, this, [this]() {
        ui->statusLabel->setText(u8"çŠ¶æ€ï¼šP2P å·²è¿æ¥");
    });
    connect(m_pcm, &PeerConnectionManager::disconnected, this, [this]() {
        ui->statusLabel->setText(u8"çŠ¶æ€ï¼šP2P å·²æ–­å¼€");
    });
    connect(m_pcm, &PeerConnectionManager::errorOccurred, this, [this](const QString& err) {
        ui->statusLabel->setText(u8"é”™è¯¯ï¼š" + err);
    });
    connect(m_pcm, &PeerConnectionManager::messageReceived, this, [this](const QString& msg) {
        // æ”¶åˆ°å¯¹æ–¹æ¶ˆæ¯ï¼Œæ˜¾ç¤ºä¸ºâ€œè¿œç«¯â€
        appendRemoteMessage(u8"å¯¹æ–¹", msg);
    });


    // ====== åˆå§‹ç•Œé¢ ======
    ui->dockChat->hide(); // åˆå§‹éšè—èŠå¤©é¢æ¿
    ui->statusLabel->setText("çŠ¶æ€ï¼šæœªè¿æ¥");

    // ====== åº•éƒ¨æ§åˆ¶æ ï¼šåŠ¨æ€æ‰©å±•æŒ‰é’®ï¼ˆä»¿è…¾è®¯ä¼šè®®ï¼‰ ======
    // ä¾æ¬¡æ·»åŠ ï¼šæ‘„åƒå¤´ã€å‚ä¼šè€…ã€å½•åˆ¶ã€ä¸¾æ‰‹ã€è®¾å¤‡ã€ç¦»å¼€ã€ç½‘ç»œè´¨é‡
    btnShareScreen = new QPushButton(u8"ğŸ“º", this);
    btnChat = new QPushButton(u8"ğŸ’¬", this);
    btnVoice = new QPushButton(u8"ğŸ¤", this);
    btnVideo = new QPushButton(u8"ğŸ¥", this);
    btnParticipants = new QPushButton(u8"ğŸ‘¥", this);
    btnRecord = new QPushButton(u8"âº", this);
    btnRaiseHand = new QPushButton(u8"âœ‹", this);
    btnDevices = new QPushButton(u8"âš™", this);
    btnLeave = new QPushButton(u8"ğŸšª ç¦»å¼€ä¼šè®®", this);
    netLabel = new QLabel(u8"ç½‘ç»œï¼šè‰¯å¥½", this);

    // å¼€å…³å‹æŒ‰é’®
    btnVoice->setCheckable(true);
    btnVoice->setChecked(isCameraOn);
    btnShareScreen->setCheckable(true);
    btnVideo->setCheckable(true);
    btnRecord->setCheckable(true);

    // å¤–è§‚å¾®è°ƒ
    btnLeave->setStyleSheet("QPushButton{background:#ff3b30;color:white;border-radius:8px;padding:10px 18px;}"
                            "QPushButton:hover{background:#ff453a;}");
    netLabel->setStyleSheet("QLabel{color:#8e8e93;padding-left:6px;}");

    // è®¾å¤‡èœå•ï¼ˆé€‰æ‹©/è®¾ç½®ï¼‰
    btnDevices->setMenu(buildDevicesMenu());

    // æŒ‚åˆ°åº•éƒ¨å¸ƒå±€
    auto bar = ui->horizontalLayout;
    bar->addWidget(btnVoice);
    bar->addWidget(btnChat);
    bar->addWidget(btnShareScreen);
    bar->addWidget(btnVideo);
    bar->addWidget(btnParticipants);
    bar->addWidget(btnRecord);
    bar->addWidget(btnRaiseHand);
    bar->addWidget(btnDevices);
    bar->addWidget(btnLeave);
    bar->addWidget(netLabel);

    // è¿æ¥æ–°å¢æŒ‰é’®
    connect(btnChat, &QPushButton::clicked, this, &shared_screen::on_btnChat_clicked);
    connect(btnVoice, &QPushButton::clicked, this, &shared_screen::on_btnVoice_clicked);
    connect(btnShareScreen, &QPushButton::clicked, this, &shared_screen::on_btnShareScreen_clicked);
    connect(btnVideo, &QPushButton::clicked, this, &shared_screen::on_btnVideo_clicked);
    connect(btnParticipants, &QPushButton::clicked, this, &shared_screen::on_btnParticipants_clicked);
    connect(btnRecord, &QPushButton::clicked, this, &shared_screen::on_btnRecord_clicked);
    connect(btnRaiseHand, &QPushButton::clicked, this, &shared_screen::on_btnRaiseHand_clicked);
    connect(btnDevices, &QPushButton::clicked, this, &shared_screen::on_btnDevices_clicked);
    connect(btnLeave, &QPushButton::clicked, this, &shared_screen::on_btnLeave_clicked);

    // ====== æ¨¡æ‹Ÿç½‘ç»œè´¨é‡ & è¿œç«¯æ¶ˆæ¯ ======
    netTimer = new QTimer(this);
    netTimer->setInterval(4000);
    connect(netTimer, &QTimer::timeout, this, &shared_screen::on_fakeNetworkTick);
    netTimer->start();

    simMsgTimer = new QTimer(this);
    simMsgTimer->setInterval(15000);
    connect(simMsgTimer, &QTimer::timeout, this, &shared_screen::on_fakeRemoteMsg);
    simMsgTimer->start();

    // ====== å¿«æ·é”® ======
    buildShortcuts();

    // åˆå§‹å±å¹•æç¤º
    ui->screenPreview->setText(u8"å±å¹•é¢„è§ˆåŒºåŸŸ\nç‚¹å‡»â€œå…±äº«å±å¹•â€å¼€å§‹");
}

shared_screen::~shared_screen()
{
    delete ui;
}

// =============== å·²æœ‰åŠŸèƒ½ ===============
void shared_screen::on_btnVoice_clicked()
{

    isVoiceOn = !isVoiceOn;
    btnVoice->setChecked(isVoiceOn);
    if (isVoiceOn)
    {
        btnVoice->setIcon(QIcon("src/icons/å½•éŸ³_voice.png"));
        ui->statusLabel->setText(u8"éº¦å…‹é£å·²å¼€å¯");
    }
    else
    {
        btnVoice->setIcon(QIcon("src/icons/å…³é—­å½•éŸ³_voice-off.png"));
        ui->statusLabel->setText(u8"éº¦å…‹é£å·²å…³é—­");
    }
    // qDebug() << "slot:on_btnVoice_clicked called";
}

void shared_screen::on_btnShareScreen_clicked()
{

    isScreenSharing = !isScreenSharing;
    btnShareScreen->setChecked(isScreenSharing);

    if (isScreenSharing)
    {
        btnShareScreen->setText(u8"â¹");
        ui->screenPreview->setText(u8"æ­£åœ¨å…±äº«å±å¹•...");
        ui->statusLabel->setText(u8"æ­£åœ¨å…±äº«å±å¹•");
    }
    else
    {
        btnShareScreen->setText(u8"ğŸ“º");
        ui->screenPreview->setText(u8"å±å¹•é¢„è§ˆåŒºåŸŸ\nç‚¹å‡»â€œå…±äº«å±å¹•â€å¼€å§‹");
        ui->statusLabel->setText(u8"æœªå…±äº«");
    }
    // qDebug() << "slot:on_btnShareScreen_clicked called";
}

void shared_screen::on_btnChat_clicked()
{

    toggleChatPanel();
}

void shared_screen::on_btnSend_clicked()
{
    const QString text = ui->chatInput->text().trimmed();
    if (text.isEmpty())
        return;

    // æ—¶é—´æˆ³ + æˆ‘çš„æ¶ˆæ¯
    const QString ts = QDateTime::currentDateTime().toString("hh:mm");
    const QString html = QString("<div style='margin:8px 0;'>"
                                 "<span style='color:#888;font-size:11px;'>%1</span> "
                                 "<span style='color:#1a73e8;font-weight:bold;'>æˆ‘ï¼š</span>"
                                 "<span style='color:#fff;'>%2</span>"
                                 "</div>")
                             .arg(ts, text);
    ui->chatView->append(html);
    ui->chatInput->clear();
    ui->chatInput->setFocus();

    // ===== æ–°å¢ï¼šé€šè¿‡ P2P å‘é€åˆ°å¯¹æ–¹ =====
    if (m_pcm) {
        m_pcm->sendMessage(text);
    }
}


// =============== æ–°å¢åŠŸèƒ½ ===============
void shared_screen::on_btnVideo_clicked()
{
    isCameraOn = !isCameraOn;
    btnVideo->setChecked(isCameraOn);
    if (isCameraOn)
    {
        btnVideo->setText(u8"ğŸ¥");

        appendSystemMessage(u8"ä½ å·²å¼€å¯æ‘„åƒå¤´");
    }
    else
    {
        btnVideo->setText(u8"ğŸ¥");
        appendSystemMessage(u8"ä½ å·²å…³é—­æ‘„åƒå¤´");
    }
}

void shared_screen::on_btnParticipants_clicked()
{
    ensureParticipantsDock();
    if (dockParticipants->isHidden())
    {
        dockParticipants->show();
        btnParticipants->setText(u8"ğŸ‘¥âœ“");
    }
    else
    {
        dockParticipants->hide();
        btnParticipants->setText(u8"ğŸ‘¥");
    }
}

void shared_screen::on_btnRecord_clicked()
{
    isRecording = !isRecording;
    btnRecord->setChecked(isRecording);
    if (isRecording)
    {
        btnRecord->setText(u8"âº");
        ui->statusLabel->setText(u8"æ­£åœ¨å½•åˆ¶");
        appendSystemMessage(u8"ä¼šè®®å½•åˆ¶å·²å¼€å§‹ï¼ˆæ¼”ç¤ºï¼‰");
    }
    else
    {
        btnRecord->setText(u8"âº");
        ui->statusLabel->setText(u8"å½•åˆ¶ç»“æŸ");
        appendSystemMessage(u8"ä¼šè®®å½•åˆ¶å·²åœæ­¢ï¼ˆæ¼”ç¤ºï¼‰");
    }
}

void shared_screen::on_btnRaiseHand_clicked()
{
    isHandRaised = !isHandRaised;
    if (isHandRaised)
    {
        btnRaiseHand->setText(u8"âœ‹");
        appendSystemMessage(u8"ä½ ä¸¾æ‰‹äº†");
    }
    else
    {
        btnRaiseHand->setText(u8"âœ‹");
        appendSystemMessage(u8"ä½ æ”¾ä¸‹äº†æ‰‹");
    }
}

void shared_screen::on_btnDevices_clicked()
{
    // è¿™é‡Œä½¿ç”¨èœå•åŠ¨ä½œå³å¯ï¼ŒçœŸå®é¡¹ç›®å¯æ‰“å¼€è®¾ç½®å¯¹è¯æ¡†
    // æ¼”ç¤ºåœ¨ç³»ç»Ÿæ¶ˆæ¯ä¸­æç¤º
    appendSystemMessage(u8"æ‰“å¼€è®¾å¤‡è®¾ç½®ï¼ˆæ¼”ç¤ºï¼‰ï¼šåˆ‡æ¢éº¦å…‹é£/æ‰¬å£°å™¨/æ‘„åƒå¤´");
}

void shared_screen::on_btnLeave_clicked()
{
    if (QMessageBox::question(this, u8"ç¦»å¼€ä¼šè®®", u8"ç¡®å®šè¦ç¦»å¼€ä¼šè®®å—ï¼Ÿ") == QMessageBox::Yes)
    {
        close();
    }
}

// =============== æ¨¡æ‹Ÿå™¨ ===============
void shared_screen::on_fakeNetworkTick()
{
    // éšæœºç½‘ç»œè´¨é‡ï¼šä¼˜ã€è‰¯ã€ä¸­ã€å·®
    static const char *labels[] = {"ä¼˜", "è‰¯", "ä¸­", "å·®"};
    int idx = QRandomGenerator::global()->bounded(0, 4);
    QString s = u8"ç½‘ç»œï¼š" + QString::fromUtf8(labels[idx]);

    // é¢œè‰²æç¤º
    QString color = "#34c759"; // ä¼˜
    if (idx == 1)
        color = "#0a84ff"; // è‰¯
    if (idx == 2)
        color = "#ffd60a"; // ä¸­
    if (idx == 3)
        color = "#ff3b30"; // å·®
    netLabel->setStyleSheet(QString("QLabel{color:%1;padding-left:6px;}").arg(color));
    netLabel->setText(s);
}

void shared_screen::on_fakeRemoteMsg()
{
    // å¦‚æœèŠå¤©é¢æ¿å…³é—­ï¼Œæ¨¡æ‹Ÿè¿œç«¯æ¶ˆæ¯ + æœªè¯»è§’æ ‡
    appendRemoteMessage(u8"å°å¼ ", u8"æ”¶åˆ°~");
    if (!isChatVisible)
    {
        unreadCount++;
        updateChatBadge();
    }
}

// =============== ç§æœ‰å¸®åŠ©å‡½æ•° ===============
void shared_screen::toggleChatPanel()
{
    isChatVisible = !isChatVisible;
    if (isChatVisible)
    {
        ui->dockChat->show();
        btnChat->setText(u8"ğŸ’¬ âœ“");
        unreadCount = 0;
        updateChatBadge();
    }
    else
    {
        ui->dockChat->hide();
        btnChat->setText(u8"ğŸ’¬");
    }
}

void shared_screen::ensureParticipantsDock()
{
    if (dockParticipants)
        return;

    dockParticipants = new QDockWidget(u8"ğŸ‘¥ å‚ä¼šè€…", this);
    participantsList = new QListWidget(dockParticipants);
    dockParticipants->setWidget(participantsList);
    dockParticipants->setMinimumWidth(260);
    addDockWidget(Qt::RightDockWidgetArea, dockParticipants);

    // å…ˆæ”¾ä¸€äº›æ¼”ç¤ºæ•°æ®
    participantsList->addItem(u8"æˆ‘ï¼ˆä¸»æŒäººï¼‰");
    participantsList->addItem(u8"å°å¼ ");
    participantsList->addItem(u8"å°æ");

    dockParticipants->hide();
}

void shared_screen::appendSystemMessage(const QString &text)
{
    const QString ts = QDateTime::currentDateTime().toString("hh:mm");
    const QString html = QString("<div style='margin:8px 0;'>"
                                 "<span style='color:#888;font-size:11px;'>%1</span> "
                                 "<span style='color:#e6e6e6;font-weight:bold;'>ç³»ç»Ÿï¼š</span>"
                                 "<span style='color:#aaa;'>%2</span>"
                                 "</div>")
                             .arg(ts, text.toHtmlEscaped());
    ui->chatView->append(html);
}

void shared_screen::appendRemoteMessage(const QString &sender, const QString &text)
{
    const QString ts = QDateTime::currentDateTime().toString("hh:mm");
    const QString html = QString("<div style='margin:8px 0;'>"
                                 "<span style='color:#888;font-size:11px;'>%1</span> "
                                 "<span style='color:#40c4ff;font-weight:bold;'>%2ï¼š</span>"
                                 "<span style='color:#fff;'>%3</span>"
                                 "</div>")
                             .arg(ts, sender.toHtmlEscaped(), text.toHtmlEscaped());
    ui->chatView->append(html);
}

void shared_screen::updateChatBadge()
{
    if (unreadCount > 0)
    {
        btnChat->setText(QString(u8"ğŸ’¬(%1)").arg(unreadCount));
    }
    else
    {
        btnChat->setText(u8"ğŸ’¬");
    }
}

void shared_screen::buildShortcuts()
{
    // Ctrl+Dï¼šéº¦å…‹é£
    auto s1 = new QShortcut(QKeySequence("Ctrl+D"), this);
    connect(s1, &QShortcut::activated, this, &shared_screen::on_btnVoice_clicked);

    // Ctrl+Eï¼šæ‘„åƒå¤´
    auto s2 = new QShortcut(QKeySequence("Ctrl+E"), this);
    connect(s2, &QShortcut::activated, this, &shared_screen::on_btnVideo_clicked);

    // Ctrl+Sï¼šå…±äº«å±å¹•
    auto s3 = new QShortcut(QKeySequence("Ctrl+S"), this);
    connect(s3, &QShortcut::activated, this, &shared_screen::on_btnShareScreen_clicked);

    // Ctrl+Hï¼šèŠå¤©
    auto s4 = new QShortcut(QKeySequence("Ctrl+H"), this);
    connect(s4, &QShortcut::activated, this, &shared_screen::on_btnChat_clicked);

    // Ctrl+Pï¼šå‚ä¼šè€…
    auto s5 = new QShortcut(QKeySequence("Ctrl+P"), this);
    connect(s5, &QShortcut::activated, this, &shared_screen::on_btnParticipants_clicked);

    // Ctrl+Rï¼šå½•åˆ¶
    auto s6 = new QShortcut(QKeySequence("Ctrl+R"), this);
    connect(s6, &QShortcut::activated, this, &shared_screen::on_btnRecord_clicked);
}

QMenu *shared_screen::buildDevicesMenu()
{
    auto m = new QMenu(this);
    m->addAction(u8"é€‰æ‹©éº¦å…‹é£...");
    m->addAction(u8"é€‰æ‹©æ‰¬å£°å™¨...");
    m->addAction(u8"é€‰æ‹©æ‘„åƒå¤´...");
    m->addSeparator();
    m->addAction(u8"æ‰“å¼€è®¾å¤‡è®¾ç½®...");
    return m;
}

// =============== â€œæŒ‰ä½è¯´è¯â€æ¼”ç¤º ===============
void shared_screen::keyPressEvent(QKeyEvent *event)
{
    if (event->key() == Qt::Key_Space && !event->isAutoRepeat())
    {
        if (!spaceHeldPTT)
        {
            spaceHeldPTT = true;
            if (!isVoiceOn)
            {
                // ä¸´æ—¶å¼€éº¦
                on_btnVoice_clicked();
                appendSystemMessage(u8"æŒ‰ä½ç©ºæ ¼é”®ï¼šä¸´æ—¶å¼€éº¦");
            }
        }
    }
    QMainWindow::keyPressEvent(event);
}

void shared_screen::keyReleaseEvent(QKeyEvent *event)
{
    if (event->key() == Qt::Key_Space && !event->isAutoRepeat())
    {
        if (spaceHeldPTT)
        {
            spaceHeldPTT = false;
            if (isVoiceOn)
            {
                // æ¾å¼€åè¿˜åŸï¼ˆè‹¥éœ€è¦å¯åˆ¤æ–­ä¹‹å‰æ˜¯å¦æ˜¯å…³éº¦ï¼‰
                on_btnVoice_clicked();
                appendSystemMessage(u8"æ¾å¼€ç©ºæ ¼é”®ï¼šè¿˜åŸä¸ºé™éŸ³");
            }
        }
    }
    QMainWindow::keyReleaseEvent(event);
}
